[Bug]: Optimizing process failed for Iceberg format table on MinIO

### What happened?

Optimizing process failed in situation:

- Iceberg Format table
- Table without partition spec
- Use MinIO

### Affects Versions

master

### What engines are you seeing the problem on?

Optimizer

### How to reproduce

- Create an Iceberg format catalog on MiniIO
- Create a table without partition spec
- Write into the table and wait for a optimizing job

### Relevant log output

```shell
2023-10-31 19:48:44,456 ERROR [Optimizer-executor-0] [com.netease.arctic.optimizer.common.OptimizerExecutor] [] - Optimizer executor[0] executed task[OptimizingTaskId(processId:1698752922559, taskId:1)] f
ailed
java.lang.RuntimeException: software.amazon.awssdk.services.s3.model.S3Exception: Object name contains unsupported characters. (Service: S3, Status Code: 400, Request ID: 17932EC87A522F80, Extended Reques
t ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8)
        at com.netease.arctic.io.ArcticFileIOAdapter.doAs(ArcticFileIOAdapter.java:47) ~[amoro-core-0.6.0-SNAPSHOT.jar:?]
        at com.netease.arctic.io.ArcticFileIOAdapter.doAs(ArcticFileIOAdapter.java:42) ~[amoro-core-0.6.0-SNAPSHOT.jar:?]
        at com.netease.arctic.optimizing.AbstractRewriteFilesExecutor.execute(AbstractRewriteFilesExecutor.java:105) ~[amoro-core-0.6.0-SNAPSHOT.jar:?]
        at com.netease.arctic.optimizing.AbstractRewriteFilesExecutor.execute(AbstractRewriteFilesExecutor.java:61) ~[amoro-core-0.6.0-SNAPSHOT.jar:?]
        at com.netease.arctic.optimizer.common.OptimizerExecutor.executeTask(OptimizerExecutor.java:105) ~[optimizer-common-0.6.0-SNAPSHOT.jar:?]
        at com.netease.arctic.optimizer.common.OptimizerExecutor.start(OptimizerExecutor.java:34) ~[optimizer-common-0.6.0-SNAPSHOT.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_291]
Caused by: software.amazon.awssdk.services.s3.model.S3Exception: Object name contains unsupported characters. (Service: S3, Status Code: 400, Request ID: 17932EC87A522F80, Extended Request ID: dd9025bab4a
d464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8)
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156) ~[aws-xml-protocol-2.20.5.jar:?]
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:108) ~[aws-xml-protocol-2.20.5.jar:?]
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:85) ~[aws-xml-protocol-2.20.5.jar:?]
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:43) ~[aws-xml-protocol-2.20.5.jar:?]
        at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler$Crc32ValidationResponseHandler.handle(AwsSyncClientHandler.java:95) ~[aws-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$7(BaseClientHandler.java:270) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:40) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:30) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:73) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:78) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:40) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:50) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:36) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:81) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:36) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:48) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:31) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206) ~[sdk-core-2.20.5.jar:?]
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37) ~[sdk-core-2.20.5.jar:?]
```


### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
+++ b/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
@@ -51,6 +51,17 @@ public class IcebergRewriteExecutor extends AbstractRewriteFilesExecutor {
   }
 
   @Override
+  protected StructLike partition() {
+    StructLike partitionData = super.partition();
+    if (partitionData != null && partitionData.size() == 0) {
+      // Cast empty partition data to NULL to avoid creating empty partition directory.
+      return null;
+    } else {
+      return partitionData;
+    }
+  }
+
+  @Override
   protected OptimizingDataReader dataReader() {
     return new GenericCombinedIcebergDataReader(
         io,

--- a/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
+++ b/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
@@ -98,19 +98,19 @@ public class IcebergRewriteExecutor extends AbstractRewriteFilesExecutor {
         new FileWriterFactory<Record>() {
 
           @Override
-          public DataWriter newDataWriter(
+          public DataWriter<Record> newDataWriter(
               EncryptedOutputFile file, PartitionSpec spec, StructLike partition) {
             return appenderFactory.newDataWriter(file, dataFileFormat(), partition);
           }
 
           @Override
-          public EqualityDeleteWriter newEqualityDeleteWriter(
+          public EqualityDeleteWriter<Record> newEqualityDeleteWriter(
               EncryptedOutputFile file, PartitionSpec spec, StructLike partition) {
             return null;
           }
 
           @Override
-          public PositionDeleteWriter newPositionDeleteWriter(
+          public PositionDeleteWriter<Record> newPositionDeleteWriter(
               EncryptedOutputFile file, PartitionSpec spec, StructLike partition) {
             return null;
           }

--- a/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
+++ b/core/src/main/java/com/netease/arctic/optimizing/IcebergRewriteExecutor.java
@@ -50,6 +50,7 @@ public class IcebergRewriteExecutor extends AbstractRewriteFilesExecutor {
     super(input, table, structLikeCollections);
   }
 
+  // TODO We can remove this override method after upgrading Iceberg version to 1.5+.
   @Override
   protected StructLike partition() {
     StructLike partitionData = super.partition();

--- a/hive/src/main/java/com/netease/arctic/hive/optimizing/MixFormatRewriteExecutor.java
+++ b/hive/src/main/java/com/netease/arctic/hive/optimizing/MixFormatRewriteExecutor.java
@@ -73,7 +73,7 @@ public class MixFormatRewriteExecutor extends AbstractRewriteFilesExecutor {
   }
 
   public long getTransactionId(List<PrimaryKeyedFile> dataFiles) {
-    return dataFiles.stream().mapToLong(PrimaryKeyedFile::transactionId).max().getAsLong();
+    return dataFiles.stream().mapToLong(PrimaryKeyedFile::transactionId).max().orElse(0L);
   }
 
   public String baseLocation() {

